examples/example_vm/example_vm.cpp

```diff
@@ -77,130 +108,157 @@ static evmc_result execute(evmc_vm* instance,
                            const uint8_t* code,
                            size_t code_size)
 {
-    evmc_result ret = {};
-    ret.status_code = EVMC_INTERNAL_ERROR;
-    if (code_size == 0)
+    ExampleVM* vm = static_cast<ExampleVM*>(instance);
+
+    if (vm->verbose > 0)
+        std::puts("execution started\n");
+
+    int64_t gas_left = msg->gas;
+    Stack stack;
+    Memory memory;
+
+    for (size_t pc = 0; pc < code_size; pc++)
     {
-        // In case of empty code return a fancy error message.
-        const char* error = rev == EVMC_BYZANTIUM ? "Welcome to Byzantium!" : "Hello Ethereum!";
-        ret.output_data = (const uint8_t*)error;
-        ret.output_size = strlen(error);
-        ret.status_code = EVMC_FAILURE;
-        ret.gas_left = msg->gas / 10;
-        ret.release = nullptr;  // We don't need to release the constant messages.
-        return ret;
-    }
+        // Check remaining gas, assume each instruction costs 1.
+        gas_left -= 1;
+        if (gas_left < 0)
+            return evmc_make_result(EVMC_OUT_OF_GAS, 0, nullptr, 0);
 
-    example_vm* vm = static_cast<example_vm*>(instance);
+        switch (code[pc])
+        {
+        default:
+            return evmc_make_result(EVMC_UNDEFINED_INSTRUCTION, 0, nullptr, 0);
 
-    // Simulate executing by checking for some code patterns.
-    // Solidity inline assembly is used in the examples instead of EVM bytecode.
+        case 0x00:  // STOP
+            return evmc_make_result(EVMC_SUCCESS, gas_left, nullptr, 0);
 
-    // Assembly: `{ mstore(0, address()) return(0, msize()) }`.
-    const char return_address[] = "\x30\x60\x00\x52\x59\x60\x00\xf3";
+        case 0x01:  // ADD
```

Comment:
Could use the opcodes from `instructions.h`?

Reply:
Can be. Although it adds dependency.

Reply:
Done.